File: src/futurnal/cli/notifications.py
"""Notification CLI commands for Phase 2D.

Provides commands for:
- Viewing notification preferences
- Updating notification settings
- Viewing notification history
- Managing do-not-disturb schedules
"""

import json
import logging
from typing import Optional

from typer import Typer, Option, Argument

logger = logging.getLogger(__name__)

notifications_app = Typer(help="Notification preferences and history commands")


def _get_notification_service():
    """Get NotificationService singleton."""
    try:
        from futurnal.notifications.service import get_notification_service
        return get_notification_service()
    except ImportError as e:
        logger.warning(f"Notification service not available: {e}")
        return None


def _get_notification_config():
    """Get NotificationConfig singleton."""
    try:
        from futurnal.notifications.config import get_notification_config
        return get_notification_config()
    except ImportError as e:
        logger.warning(f"Notification config not available: {e}")
        return None


# ============================================================================
# Preferences Commands
# ============================================================================

@notifications_app.command("preferences")
def get_preferences(
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Get current notification preferences.

    Examples:
        futurnal notifications preferences
        futurnal notifications preferences --json
    """
    try:
        config = _get_notification_config()
        if config is None:
            raise RuntimeError("Notification system not available")

        prefs = config.preferences
        data = {
            "success": True,
            "frequency": prefs.frequency.value,
            "maxDailyNotifications": prefs.max_daily_notifications,
            "minInsightConfidence": prefs.min_insight_confidence,
            "dndSchedule": {
                "enabled": prefs.dnd_schedule.enabled,
                "startTime": prefs.dnd_schedule.start_time,
                "endTime": prefs.dnd_schedule.end_time,
                "days": prefs.dnd_schedule.days,
                "isActive": prefs.dnd_schedule.is_dnd_active(),
            },
            "channels": {
                "dashboardEnabled": prefs.channels.dashboard_enabled,
                "desktopEnabled": prefs.channels.desktop_enabled,
                "minPriorityDesktop": prefs.channels.min_priority_desktop.value,
            },
        }

        if output_json:
            print(json.dumps(data))
        else:
            print("\nNotification Preferences")
            print("-" * 40)
            print(f"Frequency: {prefs.frequency.value}")
            print(f"Max Daily: {prefs.max_daily_notifications}")
            print(f"Min Confidence: {prefs.min_insight_confidence:.0%}")
            print()
            print("Do Not Disturb:")
            dnd = prefs.dnd_schedule
            status = "ACTIVE" if dnd.is_dnd_active() else "inactive"
            print(f"  Status: {status}")
            print(f"  Hours: {dnd.start_time} - {dnd.end_time}")
            print(f"  Enabled: {dnd.enabled}")
            print()
            print("Channels:")
            print(f"  Dashboard: {'enabled' if prefs.channels.dashboard_enabled else 'disabled'}")
            print(f"  Desktop: {'enabled' if prefs.channels.desktop_enabled else 'disabled'}")

    except Exception as e:
        logger.error(f"Get preferences failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "error": str(e)}))
        else:
            print(f"Error: {e}")


@notifications_app.command("set-frequency")
def set_frequency(
    frequency: str = Argument(..., help="Frequency: immediate, hourly, daily, weekly, manual"),
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Set notification frequency.

    Examples:
        futurnal notifications set-frequency daily
        futurnal notifications set-frequency hourly --json
    """
    try:
        from futurnal.notifications.config import NotificationFrequency

        # Validate frequency
        try:
            freq_enum = NotificationFrequency(frequency)
        except ValueError:
            valid = ", ".join(f.value for f in NotificationFrequency)
            raise ValueError(f"Invalid frequency '{frequency}'. Valid: {valid}")

        config = _get_notification_config()
        if config is None:
            raise RuntimeError("Notification system not available")

        config.preferences.frequency = freq_enum
        config._save()

        if output_json:
            print(json.dumps({
                "success": True,
                "frequency": freq_enum.value,
            }))
        else:
            print(f"Notification frequency set to: {freq_enum.value}")

    except Exception as e:
        logger.error(f"Set frequency failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "error": str(e)}))
        else:
            print(f"Error: {e}")


@notifications_app.command("set-dnd")
def set_dnd(
    enabled: bool = Option(None, "--enabled/--disabled", help="Enable or disable DND"),
    start: Optional[str] = Option(None, "--start", help="Start time (HH:MM)"),
    end: Optional[str] = Option(None, "--end", help="End time (HH:MM)"),
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Set do-not-disturb schedule.

    Examples:
        futurnal notifications set-dnd --start 22:00 --end 08:00
        futurnal notifications set-dnd --disabled
        futurnal notifications set-dnd --enabled --json
    """
    try:
        config = _get_notification_config()
        if config is None:
            raise RuntimeError("Notification system not available")

        dnd = config.preferences.dnd_schedule

        if enabled is not None:
            dnd.enabled = enabled

        if start is not None:
            dnd.start_time = start

        if end is not None:
            dnd.end_time = end

        config._save()

        if output_json:
            print(json.dumps({
                "success": True,
                "dndSchedule": dnd.to_dict(),
                "isActive": dnd.is_dnd_active(),
            }))
        else:
            status = "ACTIVE" if dnd.is_dnd_active() else "inactive"
            print(f"DND schedule updated: {dnd.start_time} - {dnd.end_time}")
            print(f"Enabled: {dnd.enabled}")
            print(f"Current status: {status}")

    except Exception as e:
        logger.error(f"Set DND failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "error": str(e)}))
        else:
            print(f"Error: {e}")


@notifications_app.command("set-channels")
def set_channels(
    dashboard: Optional[bool] = Option(None, "--dashboard/--no-dashboard", help="Enable/disable dashboard"),
    desktop: Optional[bool] = Option(None, "--desktop/--no-desktop", help="Enable/disable desktop"),
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Enable or disable notification channels.

    Examples:
        futurnal notifications set-channels --desktop
        futurnal notifications set-channels --no-desktop --dashboard
    """
    try:
        config = _get_notification_config()
        if config is None:
            raise RuntimeError("Notification system not available")

        channels = config.preferences.channels

        if dashboard is not None:
            channels.dashboard_enabled = dashboard

        if desktop is not None:
            channels.desktop_enabled = desktop

        config._save()

        if output_json:
            print(json.dumps({
                "success": True,
                "channels": channels.to_dict(),
            }))
        else:
            print("Notification channels updated:")
            print(f"  Dashboard: {'enabled' if channels.dashboard_enabled else 'disabled'}")
            print(f"  Desktop: {'enabled' if channels.desktop_enabled else 'disabled'}")

    except Exception as e:
        logger.error(f"Set channels failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "error": str(e)}))
        else:
            print(f"Error: {e}")


# ============================================================================
# History Commands
# ============================================================================

@notifications_app.command("history")
def get_history(
    limit: int = Option(20, "--limit", "-n", help="Maximum notifications to return"),
    unread_only: bool = Option(False, "--unread", "-u", help="Only show unread"),
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Get notification history.

    Examples:
        futurnal notifications history
        futurnal notifications history --unread --limit 10
        futurnal notifications history --json
    """
    try:
        service = _get_notification_service()
        if service is None:
            raise RuntimeError("Notification service not available")

        if unread_only:
            notifications = [n.to_dict() for n in service.dashboard.get_unread()]
        else:
            notifications = service.get_dashboard_notifications(limit)

        if output_json:
            print(json.dumps({
                "success": True,
                "notifications": notifications,
                "totalCount": len(notifications),
                "unreadCount": service.get_unread_count(),
            }))
        else:
            unread = service.get_unread_count()
            print(f"\nNotification History ({len(notifications)} shown, {unread} unread)")
            print("-" * 60)

            for n in notifications:
                read_marker = "  " if n.get("read") else "●"
                title = n.get("title", "")[:50]
                created = n.get("created_at", "")[:10]
                print(f"{read_marker} [{created}] {title}")

    except Exception as e:
        logger.error(f"Get history failed: {e}")
        if output_json:
            print(json.dumps({
                "success": False,
                "notifications": [],
                "totalCount": 0,
                "unreadCount": 0,
                "error": str(e),
            }))
        else:
            print(f"Error: {e}")


@notifications_app.command("mark-read")
def mark_read(
    notification_id: str = Argument(..., help="Notification ID to mark as read"),
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Mark a notification as read.

    Examples:
        futurnal notifications mark-read <notification_id>
    """
    try:
        service = _get_notification_service()
        if service is None:
            raise RuntimeError("Notification service not available")

        success = service.mark_notification_read(notification_id)

        if output_json:
            print(json.dumps({
                "success": success,
                "notificationId": notification_id,
            }))
        else:
            if success:
                print(f"Marked notification {notification_id} as read")
            else:
                print(f"Notification {notification_id} not found")

    except Exception as e:
        logger.error(f"Mark read failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "error": str(e)}))
        else:
            print(f"Error: {e}")


@notifications_app.command("clear")
def clear_notifications(
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Clear all notifications.

    Examples:
        futurnal notifications clear
        futurnal notifications clear --json
    """
    try:
        service = _get_notification_service()
        if service is None:
            raise RuntimeError("Notification service not available")

        count = service.dashboard.clear_all()

        if output_json:
            print(json.dumps({
                "success": True,
                "clearedCount": count,
            }))
        else:
            print(f"Cleared {count} notifications")

    except Exception as e:
        logger.error(f"Clear failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "clearedCount": 0, "error": str(e)}))
        else:
            print(f"Error: {e}")


# ============================================================================
# Status Commands
# ============================================================================

@notifications_app.command("status")
def get_status(
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Get notification system status.

    Examples:
        futurnal notifications status
        futurnal notifications status --json
    """
    try:
        service = _get_notification_service()
        if service is None:
            raise RuntimeError("Notification service not available")

        status = service.get_status()
        status["success"] = True

        if output_json:
            print(json.dumps(status))
        else:
            print("\nNotification Status")
            print("-" * 40)
            print(f"Pending insights: {status['pending_insights']}")
            print(f"Unread notifications: {status['unread_notifications']}")
            print(f"Sent today: {status['notifications_today']}/{status['max_daily']}")
            print(f"Frequency: {status['frequency']}")
            print(f"DND active: {status['dnd_active']}")
            print(f"Desktop: {'enabled' if status['desktop_enabled'] else 'disabled'}")
            print(f"Dashboard: {'enabled' if status['dashboard_enabled'] else 'disabled'}")

    except Exception as e:
        logger.error(f"Get status failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "error": str(e)}))
        else:
            print(f"Error: {e}")


@notifications_app.command("deliver")
def deliver_pending(
    force: bool = Option(False, "--force", "-f", help="Bypass DND and frequency checks"),
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Deliver pending insight notifications.

    Examples:
        futurnal notifications deliver
        futurnal notifications deliver --force
    """
    try:
        service = _get_notification_service()
        if service is None:
            raise RuntimeError("Notification service not available")

        delivered = service.deliver_pending(force=force)

        if output_json:
            print(json.dumps({
                "success": True,
                "deliveredCount": delivered,
            }))
        else:
            print(f"Delivered {delivered} insight notifications")

    except Exception as e:
        logger.error(f"Deliver failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "deliveredCount": 0, "error": str(e)}))
        else:
            print(f"Error: {e}")

File: desktop/src-tauri/src/commands/notifications.rs
//! Notification IPC commands.
//!
//! Phase 2D: Notification System - Frontend integration for:
//! - Viewing and updating notification preferences
//! - Getting notification history
//! - Managing DND schedules

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use tauri::command;

// ============================================================================
// Preferences Types
// ============================================================================

/// Do-not-disturb schedule.
#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct DndSchedule {
    pub enabled: bool,
    pub start_time: String,
    pub end_time: String,
    pub days: Vec<i32>,
    pub is_active: bool,
}

/// Channel preferences.
#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct ChannelPreferences {
    pub dashboard_enabled: bool,
    pub desktop_enabled: bool,
    pub min_priority_desktop: String,
}

/// Notification preferences response.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NotificationPreferencesResponse {
    pub success: bool,
    pub frequency: String,
    pub max_daily_notifications: i32,
    pub min_insight_confidence: f64,
    pub dnd_schedule: DndSchedule,
    pub channels: ChannelPreferences,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Get notification preferences.
///
/// Calls: `futurnal notifications preferences --json`
#[command]
pub async fn get_notification_preferences() -> Result<NotificationPreferencesResponse, String> {
    let args = vec!["notifications", "preferences", "--json"];

    let response: NotificationPreferencesResponse = match crate::python::execute_cli(&args).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Get notification preferences CLI failed: {}", e);
            return Ok(NotificationPreferencesResponse {
                success: false,
                frequency: "daily".to_string(),
                max_daily_notifications: 10,
                min_insight_confidence: 0.6,
                dnd_schedule: DndSchedule {
                    enabled: true,
                    start_time: "22:00".to_string(),
                    end_time: "08:00".to_string(),
                    days: vec![0, 1, 2, 3, 4, 5, 6],
                    is_active: false,
                },
                channels: ChannelPreferences {
                    dashboard_enabled: true,
                    desktop_enabled: true,
                    min_priority_desktop: "high".to_string(),
                },
                error: Some(format!("Failed to get preferences: {}", e)),
            });
        }
    };

    log::info!("Retrieved notification preferences (frequency={})", response.frequency);

    Ok(response)
}

/// Request to update notification frequency.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SetFrequencyRequest {
    pub frequency: String,
}

/// Response from setting frequency.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SetFrequencyResponse {
    pub success: bool,
    pub frequency: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Set notification frequency.
///
/// Calls: `futurnal notifications set-frequency <frequency> --json`
#[command]
pub async fn set_notification_frequency(
    request: SetFrequencyRequest,
) -> Result<SetFrequencyResponse, String> {
    let args = vec!["notifications", "set-frequency", &request.frequency, "--json"];

    let response: SetFrequencyResponse = match crate::python::execute_cli(&args).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Set notification frequency CLI failed: {}", e);
            return Ok(SetFrequencyResponse {
                success: false,
                frequency: request.frequency,
                error: Some(format!("Failed to set frequency: {}", e)),
            });
        }
    };

    log::info!("Set notification frequency to: {}", response.frequency);

    Ok(response)
}

/// Request to update DND schedule.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SetDndRequest {
    pub enabled: Option<bool>,
    pub start_time: Option<String>,
    pub end_time: Option<String>,
}

/// Response from setting DND.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SetDndResponse {
    pub success: bool,
    pub dnd_schedule: Option<DndSchedule>,
    pub is_active: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Set do-not-disturb schedule.
///
/// Calls: `futurnal notifications set-dnd --json [options]`
#[command]
pub async fn set_notification_dnd(
    request: SetDndRequest,
) -> Result<SetDndResponse, String> {
    let mut args = vec!["notifications".to_string(), "set-dnd".to_string(), "--json".to_string()];

    if let Some(enabled) = request.enabled {
        if enabled {
            args.push("--enabled".to_string());
        } else {
            args.push("--disabled".to_string());
        }
    }

    if let Some(start) = &request.start_time {
        args.push("--start".to_string());
        args.push(start.clone());
    }

    if let Some(end) = &request.end_time {
        args.push("--end".to_string());
        args.push(end.clone());
    }

    let args_ref: Vec<&str> = args.iter().map(|s| s.as_str()).collect();

    let response: SetDndResponse = match crate::python::execute_cli(&args_ref).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Set notification DND CLI failed: {}", e);
            return Ok(SetDndResponse {
                success: false,
                dnd_schedule: None,
                is_active: false,
                error: Some(format!("Failed to set DND: {}", e)),
            });
        }
    };

    log::info!("Updated DND schedule (active={})", response.is_active);

    Ok(response)
}

// ============================================================================
// History Types
// ============================================================================

/// A notification in the history.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NotificationItem {
    pub notification_id: String,
    pub title: String,
    pub body: String,
    pub insight_id: Option<String>,
    pub priority: String,
    pub created_at: String,
    pub delivered_at: Option<String>,
    pub read: bool,
    pub action_url: Option<String>,
    #[serde(default)]
    pub metadata: HashMap<String, serde_json::Value>,
}

/// Response from getting notification history.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NotificationHistoryResponse {
    pub success: bool,
    pub notifications: Vec<NotificationItem>,
    pub total_count: i32,
    pub unread_count: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Get notification history.
///
/// Calls: `futurnal notifications history --json [--limit N] [--unread]`
#[command]
pub async fn get_notification_history(
    limit: Option<i32>,
    unread_only: Option<bool>,
) -> Result<NotificationHistoryResponse, String> {
    let mut args = vec!["notifications".to_string(), "history".to_string(), "--json".to_string()];

    if let Some(l) = limit {
        args.push("--limit".to_string());
        args.push(l.to_string());
    }

    if unread_only.unwrap_or(false) {
        args.push("--unread".to_string());
    }

    let args_ref: Vec<&str> = args.iter().map(|s| s.as_str()).collect();

    let response: NotificationHistoryResponse = match crate::python::execute_cli(&args_ref).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Get notification history CLI failed: {}", e);
            return Ok(NotificationHistoryResponse {
                success: false,
                notifications: vec![],
                total_count: 0,
                unread_count: 0,
                error: Some(format!("Failed to get history: {}", e)),
            });
        }
    };

    log::info!(
        "Retrieved {} notifications ({} unread)",
        response.total_count,
        response.unread_count
    );

    Ok(response)
}

/// Mark a notification as read.
///
/// Calls: `futurnal notifications mark-read <notification_id> --json`
#[command]
pub async fn mark_notification_read(
    notification_id: String,
) -> Result<bool, String> {
    let args = vec!["notifications", "mark-read", &notification_id, "--json"];

    match crate::python::execute_cli::<serde_json::Value>(&args).await {
        Ok(_) => {
            log::info!("Marked notification {} as read", notification_id);
            Ok(true)
        }
        Err(e) => {
            log::warn!("Mark notification read failed: {}", e);
            Ok(false)
        }
    }
}

/// Clear all notifications.
///
/// Calls: `futurnal notifications clear --json`
#[command]
pub async fn clear_notifications() -> Result<i32, String> {
    let args = vec!["notifications", "clear", "--json"];

    #[derive(Debug, Deserialize)]
    #[serde(rename_all = "camelCase")]
    struct ClearResponse {
        cleared_count: i32,
    }

    match crate::python::execute_cli::<ClearResponse>(&args).await {
        Ok(resp) => {
            log::info!("Cleared {} notifications", resp.cleared_count);
            Ok(resp.cleared_count)
        }
        Err(e) => {
            log::warn!("Clear notifications failed: {}", e);
            Ok(0)
        }
    }
}

// ============================================================================
// Status Types
// ============================================================================

/// Notification system status.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NotificationStatusResponse {
    pub success: bool,
    pub pending_insights: i32,
    pub unread_notifications: i32,
    pub notifications_today: i32,
    pub max_daily: i32,
    pub frequency: String,
    pub dnd_active: bool,
    pub desktop_enabled: bool,
    pub dashboard_enabled: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Get notification system status.
///
/// Calls: `futurnal notifications status --json`
#[command]
pub async fn get_notification_status() -> Result<NotificationStatusResponse, String> {
    let args = vec!["notifications", "status", "--json"];

    let response: NotificationStatusResponse = match crate::python::execute_cli(&args).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Get notification status CLI failed: {}", e);
            return Ok(NotificationStatusResponse {
                success: false,
                pending_insights: 0,
                unread_notifications: 0,
                notifications_today: 0,
                max_daily: 10,
                frequency: "daily".to_string(),
                dnd_active: false,
                desktop_enabled: true,
                dashboard_enabled: true,
                error: Some(format!("Failed to get status: {}", e)),
            });
        }
    };

    log::info!(
        "Notification status: {} pending, {} unread, DND={}",
        response.pending_insights,
        response.unread_notifications,
        response.dnd_active
    );

    Ok(response)
}

/// Deliver pending notifications.
///
/// Calls: `futurnal notifications deliver --json [--force]`
#[command]
pub async fn deliver_notifications(force: Option<bool>) -> Result<i32, String> {
    let mut args = vec!["notifications".to_string(), "deliver".to_string(), "--json".to_string()];

    if force.unwrap_or(false) {
        args.push("--force".to_string());
    }

    let args_ref: Vec<&str> = args.iter().map(|s| s.as_str()).collect();

    #[derive(Debug, Deserialize)]
    #[serde(rename_all = "camelCase")]
    struct DeliverResponse {
        delivered_count: i32,
    }

    match crate::python::execute_cli::<DeliverResponse>(&args_ref).await {
        Ok(resp) => {
            log::info!("Delivered {} notifications", resp.delivered_count);
            Ok(resp.delivered_count)
        }
        Err(e) => {
            log::warn!("Deliver notifications failed: {}", e);
            Ok(0)
        }
    }
}
